deep_qa/training/text_trainer.py

```diff
@@ -485,6 +484,20 @@ def _set_padding_lengths_from_model(self):
     # but simple models might be able to get by with just the default implementation
     ########################
 
+    def _get_instance_sorting_keys(self) -> List[str]:  # pylint: disable=no-self-use
+        """
+        If we're using dynamic padding, we want to group the instances by padding length, so that
+        we minimize the amount of padding necessary per batch.  This variable sets what exactly
+        gets sorted by.  We'll call
+        :func:`~deep_qa.data.instances.instance.IndexedInstance.get_padding_lengths()` on each
+        instance, pull out these keys, and sort by them in the order specified.  You'll want to
+        override this in your model class if you have more complex models.
+
+        The default implementation is to sort first by ``num_sentence_words``, then by
+        ``num_word_characters``.
+        """
+        return ['num_sentence_words', 'num_word_characters']
```

Comment:
On your first comment: yeah, you're totally right about that.  Good catch.  I'll fix it.

On your second comment: that's kind of the whole point of the TextTrainer subclasses.  Every model is already a TextTrainer subclass, and you have to specify what instance type you're using when you specify `_build_model`.  I don't think there's a way around this; the coupling is required, because the instance type defines your model inputs and outputs.

The alternative would be to have the model come in by composition, not inheritance (e.g., we have a parallel class called 'Model' or something that implements all of the abstract methods that are currently in `TextTrainer`, and `TextTrainer` instantiates one of those, instead of subclassing `TextTrainer`).  The trouble with that is that the TextTrainer is supposed to provide a useful API to the model itself for when you're building the model, and so there would be a ton of back-and-forth between these two separate objects (i.e., you'd have to pass the TextTrainer as an argument to the `Model` constructor, or at least to `_build_model`), which is a pretty strong argument in favor of inheritance instead of composition.